`feat`: support arm64 architecture

sannidhyaroy opened the pull request:
### Changes
- [x] Verify system architecture before installation

sannidhyaroy committed:
`feat`: support arm64 architecture

sannidhyaroy commented:
@eduardoleolim Could you check if this script properly installs the aarch64 version of Zen on a ARM machine?
```shell
bash <(curl -s https://raw.githubusercontent.com/sannidhyaroy/updates-server/refs/heads/feat/arch/install.sh)
``` 
The PR should be ready, if it's working as expected...

eduardoleolim commented:
It works

![Captura desde 2025-02-17 08-20-27](https://github.com/user-attachments/assets/3715ba0f-01d9-426a-a05b-49167a4eee15)

![Captura desde 2025-02-17 08-20-52](https://github.com/user-attachments/assets/ae1b2952-65b6-46d0-9718-5bd47f39a120)

eduardoleolim commented:
@sannidhyaroy , I noticed an error. If there is a case where the output of `uname -m` is `arm64`, so the variable `os_arch` must be updated with the value `aarch64` to build the download url correctly.

I think the download url must be built after switch statement

sannidhyaroy commented:
> @sannidhyaroy , I noticed an error. If there is a case where the output of `uname -m` is `arm64`, so the variable `os_arch` must be updated with the value `aarch64` to build the download url correctly.
> 
> I think the download url must be built after switch statement

Right, forgot about that completely. Thanks, I'll push a fix soon!

sannidhyaroy committed:
`fix`: set proper download url for amd64

sannidhyaroy commented:
> @sannidhyaroy , I noticed an error. If there is a case where the output of `uname -m` is `arm64`

That should work now.

eduardoleolim commented:
It works but I think it could be better

```bash
set -euo pipefail

os_arch=$(uname -m)
... Variables initialized here
# Delete initialization of official_package_location in line 10
... Other variables initialized here

...

case "$os_arch" in
    x86_64) echo "64-bit (Intel/AMD) architecture identified!" ;;
    aarch64|arm64)
		echo "64-bit ARM architecture identified!"
		os_arch="aarch64" # Update os_arch to ensure valid value for download url. Another option could be extract arm64 to another case of switch statement
    *)
		echo "Zen doesn't support this architecture: $os_arch"
		exit 1 ;;
esac

# Make the initialization of official_package_location after define the architecture correctly
official_package_location="https://github.com/zen-browser/desktop/releases/latest/download/zen.linux-$os_arch.tar.xz"

echo "Downloading the latest package"
curl -L --progress-bar -o $tar_location $official_package_location

...

```
Three changes was made: 

1. Delete initialization of official_package_location in line 10
2. Update os_arch in switch statement
3. Initialize official_package_location after switch statement

I think this make the script a little bit smaller and better. I add the full script.

[zen.txt](https://github.com/user-attachments/files/18864392/zen.txt)

sannidhyaroy commented:
That's what I originally did. But then scraped it off, thinking that all variables being at the top makes it easier for others to make changes, if required, since you wouldn't need to check through the entire script to find the variable declaration.

I can change the code, but what do you think about this?

mr-cheffy committed:
Merge branch 'main' into feat/arch

eduardoleolim commented:
I made another alternative:

1. I added an if statement next to os_arch initialization to update it from **arm64** to **aarch64**. I consider the if statement as part of variables initialization. 
2. I removed the case **arm64** from switch statement.
3. The initialization of official_package_location keeps in line 10 (now 13).

[zen.txt](https://github.com/user-attachments/files/18873210/zen.txt)

sannidhyaroy commented on install.sh:
```diff
@@ -2,11 +2,12 @@
 universal_path_for_installation_directory="$HOME/$literal_name_of_installation_directory"
 app_installation_directory="$universal_path_for_installation_directory/zen"
-official_package_location="https://github.com/zen-browser/desktop/releases/latest/download/zen.linux-x86_64.tar.xz"
+official_package_location="https://github.com/zen-browser/desktop/releases/latest/download/zen.linux-$os_arch.tar.xz"
```
I think we can declare the variable here, but leave it's value as an empty placeholder, then go ahead with the original solution. That way all the variables will be declared at the top, so the code is more maintainable for future developers.

sannidhyaroy commented on install.sh:
```diff
@@ -16,12 +17,22 @@ desktop_in_local_applications="$local_application_path/$app_name.desktop"
+    x86_64) echo "64-bit (Intel/AMD) architecture identified!" ;;
+    aarch64|arm64)
+		echo "64-bit ARM architecture identified!"
+		official_package_location="https://github.com/zen-browser/desktop/releases/latest/download/zen.linux-aarch64.tar.xz" ;;
```
Set the `os-arch` value inside the case

sannidhyaroy commented on install.sh:
```diff
@@ -16,12 +17,22 @@ desktop_in_local_applications="$local_application_path/$app_name.desktop"
+		echo "Zen doesn't support this architecture: $os_arch"
+		exit 1 ;;
+esac
+
```
And set our `official_package_location` value outside the case

sannidhyaroy reviewed (commented):
This has the pros of the above mentioned reasons that you specified, while handling the cons that I mentioned to some extent.

```shell
#!/bin/bash

set -euo pipefail

# Variables
...
official_package_location="" # Placeholder for download URL, to be set later
...

# Handle architecture detection and set the correct download URL
case "$os_arch" in
    x86_64) 
        echo "64-bit (Intel/AMD) architecture identified!"
        ;;
    aarch64|arm64) 
        echo "64-bit ARM architecture identified!"
        os_arch="aarch64"
        ;;
    *)
        echo "Zen does not support this architecture: $os_arch"
        exit 1
        ;;
esac

# Set the official package download URL after architecture is set
official_package_location="https://github.com/zen-browser/desktop/releases/latest/download/zen.linux-$os_arch.tar.xz"

# Rest of the script
...
``` 

I think this is the midway between both the solutions. What do you think?

eduardoleolim commented:
I like it

sannidhyaroy commented:
> I like it

Great! We'll proceed with that approach then...

sannidhyaroy committed:
`refactor(architecture)`: download url logic for arm architectures

sannidhyaroy committed:
`fix`: verify os before installation

The script will fail on non-linux systems to avoid accidental execution on unsupported platforms

eduardoleolim commented on install.sh:
```diff
@@ -17,6 +17,13 @@ desktop_in_local_applications="$local_application_path/$app_name.desktop"
+# Check OS
+if [[ "$(uname)" != "Linux" ]]; then
+    echo "This script is only for Linux."
+    echo "Visit https://github.com/zen-browser/desktop#-installation to learn more about supported operating systems"
```
@sannidhyaroy Could you update the link for `https://github.com/zen-browser/desktop#other-linux-distributions-tarball-or-appimage`?

eduardoleolim commented:
@mauro-balades do you see any bug?
